proc/traverse/over.go

```diff
@@ -1,85 +1,91 @@
 package traverse
 
 import (
+	"context"
+
 	"github.com/brimdata/zed"
 	"github.com/brimdata/zed/expr"
 	"github.com/brimdata/zed/proc"
 	"github.com/brimdata/zed/zbuf"
 )
 
 type Over struct {
+	parent zbuf.Puller
 	exprs  []expr.Evaluator
-	parent proc.Interface
+	outer  []zed.Value
 	batch  zbuf.Batch
-	vals   []zed.Value
-	eof    bool
+	enter  *Enter
 }
 
-func NewOver(parent proc.Interface, exprs []expr.Evaluator) *Over {
+func NewOver(pctx *proc.Context, parent zbuf.Puller, exprs []expr.Evaluator) *Over {
 	return &Over{
-		exprs:  exprs,
 		parent: parent,
+		exprs:  exprs,
 	}
 }
 
-func (o *Over) Pull() (zbuf.Batch, error) {
-	if len(o.vals) == 0 {
-		batch, err := o.parent.Pull()
+func (o *Over) AddScope(ctx context.Context, names []string, exprs []expr.Evaluator) *Scope {
+	scope := newScope(ctx, o, names, exprs)
+	o.enter = scope.enter
+	return scope
+}
+
+func (o *Over) Pull(done bool) (zbuf.Batch, error) {
+	if len(o.outer) == 0 {
+		batch, err := o.parent.Pull(done)
 		if batch == nil || err != nil {
-			return batch, err
+			return nil, err
 		}
-		o.eof = false
 		o.batch = batch
-		o.vals = batch.Values()
+		o.outer = batch.Values()
 	}
-	if o.eof {
-		o.eof = false
+	if done {
+		o.outer = nil
 		return nil, nil
 	}
```

Comment:
When called with `done` equal to true, shouldn't `Pull` always call `o.parent.Pull(true)`?

My understanding is that it should, so I was expecting this done/outer prelude to look like this:
```go
	if done {
		o.outer = nil
		return o.parent.Pull(true)
	}
	if len(o.outer) == 0 {
		batch, err := o.parent.Pull(false)
		if batch == nil || err != nil {
			return nil, err
		}
		o.batch = batch
		o.outer = batch.Values()
	}
```